{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL\n",
      "Or type server to fetch the genome. CENpk\n",
      "\n",
      "Hold tight while I fetch that genome...\n",
      "\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'CENPK113-7D_CH01.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-2451df652d8d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 173\u001b[0;31m \u001b[0mfindCutSites\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    174\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-1-2451df652d8d>\u001b[0m in \u001b[0;36mfindCutSites\u001b[0;34m()\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mspecies\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"CENpk\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m         allRecords=[SeqIO.read(\"CENPK113-7D_CH01.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH02.fasta\", \"fasta\"),\n\u001b[0m\u001b[1;32m     65\u001b[0m                     \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Scer03.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Scer04.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m                     \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Scer05.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Scer06.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    653\u001b[0m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0mfirst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m         \u001b[0mfirst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    570\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid alphabet, %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    571\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 572\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mas_handle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    573\u001b[0m         \u001b[0;31m# Map the file format to a sequence iterator:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    574\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/contextlib.py\u001b[0m in \u001b[0;36m__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__enter__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 59\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"generator didn't yield\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/Bio/File.py\u001b[0m in \u001b[0;36mas_handle\u001b[0;34m(handleish, mode, **kwargs)\u001b[0m\n\u001b[1;32m     88\u001b[0m                 \u001b[0;32myield\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 90\u001b[0;31m             \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandleish\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     91\u001b[0m                 \u001b[0;32myield\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     92\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'CENPK113-7D_CH01.fasta'"
     ]
    }
   ],
   "source": [
    "'''\n",
    "cutFinder\n",
    "by Leo d'Espaux <leodespaux@gmail.com>\n",
    "with help from William Zhuang, Kai Li\n",
    "\n",
    "Finds cut sites within a user-input sequence, then checks those candidates against\n",
    "chromosome files discarding sequences found more than once. \n",
    "Output is a list of candidate 23mers in your target sequence.\n",
    "\n",
    "Details:\n",
    "*For now only looks in the sense strand, but checks both sense and antisense on the genome.\n",
    "*The program lists these sequences as it finds them. \n",
    "*Usually, we get a cut site every ~25nt.\n",
    "*Note your target can be a sequence that's not originally in the genome.\n",
    "\n",
    "Current as of 9/1/15\n",
    "\n",
    "'''\n",
    "\n",
    "# import libraries\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def findCutSites():\n",
    "\n",
    "    print(\"I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL\")\n",
    "    species=input(\"Or type server to fetch the genome. \")\n",
    "    print(\"\")\n",
    "    print(\"Hold tight while I fetch that genome...\")\n",
    "    print(\"\")\n",
    "    \n",
    "    if species==\"server\":\n",
    "        allRecords=fetchNCBI()\n",
    "        \n",
    "    elif species==\"PO1f\":\n",
    "        allRecords=[SeqIO.read(\"Yl_PO1f_A.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_B.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_C.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_D.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_E.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_F.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"CLIB\":\n",
    "        allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"SAKL\":\n",
    "        allRecords=[SeqIO.read(\"chromosomeA.fasta\", \"fasta\"), SeqIO.read(\"chromosomeB.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeC.fasta\", \"fasta\"), SeqIO.read(\"chromosomeD.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeE.fasta\", \"fasta\"), SeqIO.read(\"chromosomeF.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"S288c\":\n",
    "        allRecords=[SeqIO.read(\"Scer01.fasta\", \"fasta\"), SeqIO.read(\"Scer02.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer03.fasta\", \"fasta\"), SeqIO.read(\"Scer04.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer05.fasta\", \"fasta\"), SeqIO.read(\"Scer06.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer07.fasta\", \"fasta\"), SeqIO.read(\"Scer08.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer09.fasta\", \"fasta\"), SeqIO.read(\"Scer10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer11.fasta\", \"fasta\"), SeqIO.read(\"Scer12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer13.fasta\", \"fasta\"), SeqIO.read(\"Scer14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer15.fasta\", \"fasta\"), SeqIO.read(\"Scer16.fasta\", \"fasta\"),]\n",
    "        \n",
    "    elif species==\"CENpk\":\n",
    "        allRecords=[SeqIO.read(\"CENPK113-7D_CH1.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH2.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH3.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH4.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH5.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH6.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH7.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH8.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH9.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH11.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH13.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH15.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH16.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_mitochondria.fasta\", \"fasta\")]\n",
    "        \n",
    "    print(\"I found the following genomic records: \")\n",
    "    for record in allRecords:\n",
    "        print( record.description)\n",
    "    print(\"\")\n",
    "\n",
    "\n",
    "    # ask for target and see if it's present and unique\n",
    "    print(\"OK, what sequence do you want to find cut sites in: \")\n",
    "    targetSeq=Seq(input().replace(\" \",\"\"))\n",
    "    \n",
    "    print(\"\")\n",
    "    targetCount=checkUnique(allRecords, targetSeq)\n",
    "    print(\"Found instances: \")\n",
    "    print(targetCount)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #some sensible checks\n",
    "    if targetCount==0:\n",
    "        print(\"Sorry your sequence is not found in this genome\")\n",
    "        exit()\n",
    "    elif targetCount>1:\n",
    "        print(\"Sorry there are more than one instances of your sequence in this genome.\")\n",
    "        exit()\n",
    "    \n",
    "    elif targetCount==1:\n",
    "        print(\"Your sequence is found in the genome once, proceeding...\")\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Here are some suitable cut sites for your target\")\n",
    "\n",
    "    #make an empty list\n",
    "    cutters=[]    \n",
    "    \n",
    "    # let's look through our locus to find cut sites. We end at len-24, since we are looking thru 23-nts,  \n",
    "    # and because python includes the first index but not the last one\n",
    "    uniqueCount=0\n",
    "    degenerateCount=0\n",
    "    NAGcount=0 \n",
    "    \n",
    "    for i in range(len(targetSeq)-24):\n",
    "\n",
    "        # sub is an array containing all 23mers\n",
    "        sub=targetSeq[i:i+23]\n",
    "        \n",
    "        # first, cut sequences are N20NGG, cannot contain TTTTTT\n",
    "        if sub[21]==\"G\" and sub[22]=='G' and not('TTTTTT' in sub):\n",
    "           \n",
    "            # and also, if the last 12 bases exists followed by any NGG more than once, toss it.\n",
    "            # we define a CheckUnique function below which returns 1 if the input is found only once,\n",
    "            # returns 2 if it's found more than once, and 0 if it's not found at all. \n",
    "        \n",
    "            if (checkUnique(allRecords,sub[8:20]+\"AGG\")+checkUnique(allRecords,sub[8:20]+\"TGG\")+checkUnique(allRecords,sub[8:20]+\"CGG\")+checkUnique(allRecords,sub[8:20]+\"GGG\")) > 1:\n",
    "                degenerateCount=degenerateCount+1\n",
    "                \n",
    "            \n",
    "            # also, don't want any N9-20 followed by NAG\n",
    "            elif (checkUnique(allRecords,sub[8:20]+\"AAG\")+checkUnique(allRecords,sub[8:20]+\"TAG\")+checkUnique(allRecords,sub[8:20]+\"CAG\")+checkUnique(allRecords,sub[8:20]+\"GAG\")) > 0:\n",
    "                NAGcount=NAGcount+1\n",
    "                \n",
    "            \n",
    "            else:\n",
    "                # print the cut site that passes all tests\n",
    "                uniqueCount=uniqueCount+1\n",
    "                print (sub[0:20])\n",
    "    print(\"\\n\"+\"Done finding sequences. Good luck!\")  \n",
    "    print(\"degenerate, NAG counts: \")\n",
    "    print(degenerateCount)\n",
    "    print(NAGcount)\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "            \n",
    "            \n",
    "            \n",
    "def checkUnique(searchRecords, string):\n",
    "# Here we take in a string sequence and see if it's unique in the records (chromosomes)\n",
    "# defined earlier. If unique, returns 1; if not found, 0; and if found more than once, 2.\n",
    "        \n",
    "        nfound=0\n",
    "        \n",
    "        for record in searchRecords:\n",
    "            nfound=nfound+record.seq.count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            # note that we should look in the other strand, too    \n",
    "            nfound=nfound+record.seq.reverse_complement().count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            \n",
    "        if nfound == 0:\n",
    "            return 0\n",
    "            \n",
    "        if nfound == 1:\n",
    "            return 1\n",
    "            \n",
    "            \n",
    "findCutSites()\n",
    "                    \n",
    "                        \n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
