{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What species do you want, e.g. PO1f, s288c\n",
      "Or type server to fetch the genome. s288c\n",
      "\n",
      "Hold tight while I fetch that genome...\n",
      "\n",
      "I found the following genomic records: \n",
      "tpg|BK006935.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=I] [note=R64-1-1]\n",
      "tpg|BK006936.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=II] [note=R64-1-1]\n",
      "tpg|BK006937.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=III] [note=R64-1-1]\n",
      "tpg|BK006938.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IV] [note=R64-1-1]\n",
      "tpg|BK006939.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=V] [note=R64-1-1]\n",
      "tpg|BK006940.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VI] [note=R64-1-1]\n",
      "tpg|BK006941.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VII] [note=R64-1-1]\n",
      "tpg|BK006934.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VIII] [note=R64-1-1]\n",
      "tpg|BK006942.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IX] [note=R64-1-1]\n",
      "tpg|BK006943.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=X] [note=R64-1-1]\n",
      "tpg|BK006944.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XI] [note=R64-1-1]\n",
      "tpg|BK006945.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XII] [note=R64-1-1]\n",
      "tpg|BK006946.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIII] [note=R64-1-1]\n",
      "tpg|BK006947.3| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIV] [note=R64-1-1]\n",
      "tpg|BK006948.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XV] [note=R64-1-1]\n",
      "tpg|BK006949.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XVI] [note=R64-1-1]\n",
      "\n",
      "OK, what sequence do you want to find cut sites in: \n",
      "TGTTGGAATAGAAATCAACTATCATCTACTAACTAGTATTTACATTACTAGTATATTATCATATACGGTGTTAGAAGATGACGCAAATGATGAGAAATAGTCATCTAAATTAGTGGAAGCTGAAACGCAAGGATTGATAATGTAATAGGATCAATGAATATAAACATATAAAATGATGATAATAATATTTATAGAATTGTGTAGAATTGCAGATTCCCTTTTATGGATTCCTAAATCCTTGAGGAGAACTTCTAGTATATTCTGTATACCTAATATTATAGCCTTTATCAACAATGGAATCCCAACAATTATCTCAACATTCAC\n",
      "\n",
      "Found instances: \n",
      "2\n",
      "Sorry there are more than one instances of your sequence in this genome.\n",
      "\n",
      "Here are some suitable cut sites for your target\n",
      "\n",
      "Done finding sequences. Good luck!\n",
      "degenerate, NAG counts: \n",
      "7\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "cutFinder\n",
    "by Leo d'Espaux and William Zhuang\n",
    "\n",
    "Finds cut sites within a user-input sequence, having the chromosome files saved locally\n",
    "or retrieving from NCBI using entrez.\n",
    "\n",
    "Details:\n",
    "*For now only looks in the sense strand, but checks both sense and antisense on the genome.\n",
    "*The program lists these sequences as it finds them. \n",
    "*Usually, we get a cut site every ~25nt.\n",
    "\n",
    "Current as of 9/1/15\n",
    "\n",
    "'''\n",
    "\n",
    "# import libraries\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def findCutSites():\n",
    "\n",
    "    print(\"What species do you want, e.g. PO1f, s288c\")\n",
    "    species=input(\"Or type server to fetch the genome. \")\n",
    "    print(\"\")\n",
    "    print(\"Hold tight while I fetch that genome...\")\n",
    "    print(\"\")\n",
    "    \n",
    "    if species==\"server\":\n",
    "        allRecords=fetchNCBI()\n",
    "        \n",
    "    elif species==\"PO1f\":\n",
    "        allRecords=[SeqIO.read(\"Yl_PO1f_A.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_B.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_C.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_D.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_E.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_F.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"CLIB\":\n",
    "        allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"SAKL\":\n",
    "        allRecords=[SeqIO.read(\"chromosomeA.fasta\", \"fasta\"), SeqIO.read(\"chromosomeB.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeC.fasta\", \"fasta\"), SeqIO.read(\"chromosomeD.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeE.fasta\", \"fasta\"), SeqIO.read(\"chromosomeF.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"s288c\":\n",
    "        allRecords=[SeqIO.read(\"Scer01.fasta\", \"fasta\"), SeqIO.read(\"Scer02.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer03.fasta\", \"fasta\"), SeqIO.read(\"Scer04.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer05.fasta\", \"fasta\"), SeqIO.read(\"Scer06.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer07.fasta\", \"fasta\"), SeqIO.read(\"Scer08.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer09.fasta\", \"fasta\"), SeqIO.read(\"Scer10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer11.fasta\", \"fasta\"), SeqIO.read(\"Scer12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer13.fasta\", \"fasta\"), SeqIO.read(\"Scer14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer15.fasta\", \"fasta\"), SeqIO.read(\"Scer16.fasta\", \"fasta\"),]\n",
    "        \n",
    "    print(\"I found the following genomic records: \")\n",
    "    for record in allRecords:\n",
    "        print( record.description)\n",
    "    print(\"\")\n",
    "\n",
    "\n",
    "    # ask for target and see if it's present and unique\n",
    "    print(\"OK, what sequence do you want to find cut sites in: \")\n",
    "    targetSeq=Seq(input().replace(\" \",\"\"))\n",
    "    \n",
    "    print(\"\")\n",
    "    targetCount=checkUnique(allRecords, targetSeq)\n",
    "    print(\"Found instances: \")\n",
    "    print(targetCount)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #some sensible checks\n",
    "    if targetCount==0:\n",
    "        print(\"Sorry your sequence is not found in this genome\")\n",
    "        exit()\n",
    "    elif targetCount>1:\n",
    "        print(\"Sorry there are more than one instances of your sequence in this genome.\")\n",
    "        exit()\n",
    "    \n",
    "    elif targetCount==1:\n",
    "        print(\"Your sequence is found in the genome once, proceeding...\")\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Here are some suitable cut sites for your target\")\n",
    "\n",
    "    #make an empty list\n",
    "    cutters=[]    \n",
    "    \n",
    "    # let's look through our locus to find cut sites. We end at len-24, since we are looking thru 23-nts,  \n",
    "    # and because python includes the first index but not the last one\n",
    "    uniqueCount=0\n",
    "    degenerateCount=0\n",
    "    NAGcount=0 \n",
    "    \n",
    "    for i in range(len(targetSeq)-24):\n",
    "\n",
    "        # sub is an array containing all 23mers\n",
    "        sub=targetSeq[i:i+23]\n",
    "        \n",
    "        # first, cut sequences are N20NGG, cannot contain TTTTTT\n",
    "        if sub[21]==\"G\" and sub[22]=='G' and not('TTTTTT' in sub):\n",
    "           \n",
    "            # and also, if the last 12 bases exists followed by any NGG more than once, toss it.\n",
    "            # we define a CheckUnique function below which returns 1 if the input is found only once,\n",
    "            # returns 2 if it's found more than once, and 0 if it's not found at all. \n",
    "        \n",
    "            if (checkUnique(allRecords,sub[8:20]+\"AGG\")+checkUnique(allRecords,sub[8:20]+\"TGG\")+checkUnique(allRecords,sub[8:20]+\"CGG\")+checkUnique(allRecords,sub[8:20]+\"GGG\")) > 1:\n",
    "                degenerateCount=degenerateCount+1\n",
    "                \n",
    "            \n",
    "            # also, don't want any N9-20 followed by NAG\n",
    "            elif (checkUnique(allRecords,sub[8:20]+\"AAG\")+checkUnique(allRecords,sub[8:20]+\"TAG\")+checkUnique(allRecords,sub[8:20]+\"CAG\")+checkUnique(allRecords,sub[8:20]+\"GAG\")) > 0:\n",
    "                NAGcount=NAGcount+1\n",
    "                \n",
    "            \n",
    "            else:\n",
    "                # print the cut site that passes all tests\n",
    "                uniqueCount=uniqueCount+1\n",
    "                print (sub[0:20])\n",
    "    print(\"\\n\"+\"Done finding sequences. Good luck!\")  \n",
    "    print(\"degenerate, NAG counts: \")\n",
    "    print(degenerateCount)\n",
    "    print(NAGcount)\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "            \n",
    "            \n",
    "            \n",
    "def checkUnique(searchRecords, string):\n",
    "# Here we take in a string sequence and see if it's unique in the records (chromosomes)\n",
    "# defined earlier. If unique, returns 1; if not found, 0; and if found more than once, 2.\n",
    "        \n",
    "        nfound=0\n",
    "        \n",
    "        for record in searchRecords:\n",
    "            nfound=nfound+record.seq.count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            # note that we should look in the other strand, too    \n",
    "            nfound=nfound+record.seq.reverse_complement().count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            \n",
    "        if nfound == 0:\n",
    "            return 0\n",
    "            \n",
    "        if nfound == 1:\n",
    "            return 1\n",
    "            \n",
    "            \n",
    "findCutSites()\n",
    "                    \n",
    "                        \n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
