{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL\n",
      "Or type server to fetch the genome. CENpk\n",
      "\n",
      "Hold tight while I fetch that genome...\n",
      "\n",
      "I found the following genomic records: \n",
      "@gi|144228165:1-230208\n",
      "@gi|50593115:1-813178\n",
      "@gi|85666111:1-316617\n",
      "@gi|93117368:1-1531918\n",
      "@gi|7276232:1-576869\n",
      "@gi|42742172:1-270148\n",
      "@gi|83578099:1-1090946\n",
      "@gi|82795252:1-562643\n",
      "@gi|6322016:1-439885\n",
      "@gi|116006492:1-745745\n",
      "@gi|83722562:1-666454\n",
      "@gi|85666119:1-1078175\n",
      "@gi|44829554:1-924429\n",
      "@gi|117937805:1-784333\n",
      "@gi|84626310:1-1091289\n",
      "@gi|50593503:1-948062\n",
      "@gi|6226515:1-85779\n",
      "\n",
      "OK, what sequence do you want to find cut sites in: \n",
      "AAGATAAGTTGTATCCATTCACTTGGGATGCCGTCAGATACAACGGTAAGTTAATCGCCTACCCAATTGCTGTTGAAGCTTTGTCTTTGATCTACAATAAGGACTTGTTACCTAACCCACCAAAGACCTGGGAAGAAATCCCAGCTTTAGATAAGGAGTTAAAAGCTAAGGGTAAGTCCGCTTTGATGTTTAACTTGCAAGAACCATACTTCACTTGGCCATTGATCGCTGCTGATGGTGGTTACGCTTTTAAGTATGAAAACGGTAAATACGACATTAAGGATGTCGGTGTCGACAATGCTGGTGCTAAGGCCGGTTTAACTTTCTTAGTCGATTTGATTAAGAATAAACATATGAATGCTGACACTGATTACTCTATTGCTGAAGCTGCTTTCAACAAGGGTGAAACCGCTATGACTATTAACGGTCCATGGGCCTGGTCTAACATTGATACCTCTAAAGTCAACTACGGTGTCACCGTCTTGCCAACTTTTAAGGGTCAACCATCTAAGCCATTCGTCGGTGTCTTGTCTGCCGGTATTAACGCTGCCTCTCCAAATAAGGAATTGGCCAAGGAATTCTTAGAAAACTACTTGTTAACCGATGAAGGTTTAGAGGCCGTTAACAAGGATAAGCCATTAGGTGCTGTTGCTTTGAAGTCTTACGAAGAAGAGTTGGCTAAGGATCCAAGAATTGCTGCTACTATGGAAAACGCTCAAAA\n",
      "\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Mixed RNA/DNA found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-ebe000e82abb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 171\u001b[0;31m \u001b[0mfindCutSites\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    172\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-1-ebe000e82abb>\u001b[0m in \u001b[0;36mfindCutSites\u001b[0;34m()\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 82\u001b[0;31m     \u001b[0mtargetCount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcheckUnique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mallRecords\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtargetSeq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     83\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Found instances: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtargetCount\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-1-ebe000e82abb>\u001b[0m in \u001b[0;36mcheckUnique\u001b[0;34m(searchRecords, string)\u001b[0m\n\u001b[1;32m    157\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    158\u001b[0m             \u001b[0;31m# note that we should look in the other strand, too\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 159\u001b[0;31m             \u001b[0mnfound\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnfound\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreverse_complement\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstring\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    160\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mnfound\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/Bio/Seq.py\u001b[0m in \u001b[0;36mreverse_complement\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    809\u001b[0m         \"\"\"\n\u001b[1;32m    810\u001b[0m         \u001b[0;31m# Use -1 stride/step to reverse the complement\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 811\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomplement\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    812\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    813\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mtranscribe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/leodespaux/anaconda/lib/python3.4/site-packages/Bio/Seq.py\u001b[0m in \u001b[0;36mcomplement\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    767\u001b[0m         \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'T'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m't'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    768\u001b[0m             \u001b[0;31m# TODO - Handle this cleanly?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 769\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Mixed RNA/DNA found\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    770\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0;34m'U'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m'u'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    771\u001b[0m             \u001b[0mttable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_rna_complement_table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Mixed RNA/DNA found"
     ]
    }
   ],
   "source": [
    "'''\n",
    "cutFinder\n",
    "by Leo d'Espaux <leodespaux@gmail.com>\n",
    "with help from William Zhuang, Kai Li\n",
    "\n",
    "Finds cut sites within a user-input sequence, then checks those candidates against\n",
    "chromosome files discarding sequences found more than once. \n",
    "Output is a list of candidate 23mers in your target sequence.\n",
    "\n",
    "Details:\n",
    "*For now only looks in the sense strand, but checks both sense and antisense on the genome.\n",
    "*The program lists these sequences as it finds them. \n",
    "*Usually, we get a cut site every ~25nt.\n",
    "*Note your target can be a sequence that's not originally in the genome.\n",
    "\n",
    "Current as of 9/1/15\n",
    "\n",
    "'''\n",
    "\n",
    "# import libraries\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def findCutSites():\n",
    "\n",
    "    print(\"I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL\")\n",
    "    species=input(\"Or type server to fetch the genome. \")\n",
    "    print(\"\")\n",
    "    print(\"Hold tight while I fetch that genome...\")\n",
    "    print(\"\")\n",
    "    \n",
    "    if species==\"PO1f\":\n",
    "        allRecords=[SeqIO.read(\"Yl_PO1f_A.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_B.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_C.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_D.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_E.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_F.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"CLIB\":\n",
    "        allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"SAKL\":\n",
    "        allRecords=[SeqIO.read(\"chromosomeA.fasta\", \"fasta\"), SeqIO.read(\"chromosomeB.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeC.fasta\", \"fasta\"), SeqIO.read(\"chromosomeD.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeE.fasta\", \"fasta\"), SeqIO.read(\"chromosomeF.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"S288c\":\n",
    "        allRecords=[SeqIO.read(\"Scer01.fasta\", \"fasta\"), SeqIO.read(\"Scer02.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer03.fasta\", \"fasta\"), SeqIO.read(\"Scer04.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer05.fasta\", \"fasta\"), SeqIO.read(\"Scer06.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer07.fasta\", \"fasta\"), SeqIO.read(\"Scer08.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer09.fasta\", \"fasta\"), SeqIO.read(\"Scer10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer11.fasta\", \"fasta\"), SeqIO.read(\"Scer12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer13.fasta\", \"fasta\"), SeqIO.read(\"Scer14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer15.fasta\", \"fasta\"), SeqIO.read(\"Scer16.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"CENpk\":\n",
    "        allRecords=[SeqIO.read(\"CENPK113-7D_CH1.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH2.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH3.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH4.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH5.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH6.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH7.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH8.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH9.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH11.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH13.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH15.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH16.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_mitochondria.fasta\", \"fasta\")]\n",
    "        \n",
    "    print(\"I found the following genomic records: \")\n",
    "    for record in allRecords:\n",
    "        print( record.description)\n",
    "    print(\"\")\n",
    "\n",
    "\n",
    "    # ask for target and see if it's present and unique\n",
    "    print(\"OK, what sequence do you want to find cut sites in: \")\n",
    "    targetSeq=Seq(input().replace(\" \",\"\").upper()) #get rid of spaces and make all caps\n",
    "    \n",
    "    print(\"\")\n",
    "    targetCount=checkUnique(allRecords, targetSeq)\n",
    "    print(\"Found instances: \")\n",
    "    print(targetCount)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #some sensible checks\n",
    "    if targetCount==0:\n",
    "        print(\"Sorry your sequence is not found in this genome\")\n",
    "        exit()\n",
    "    elif targetCount>1:\n",
    "        print(\"Sorry there are more than one instances of your sequence in this genome.\")\n",
    "        exit()\n",
    "    \n",
    "    elif targetCount==1:\n",
    "        print(\"Your sequence is found in the genome once, proceeding...\")\n",
    "    \n",
    "    print(\"\")\n",
    "    print(\"Here are some suitable cut sites for your target\")\n",
    "\n",
    "    #make an empty list\n",
    "    cutters=[]    \n",
    "    \n",
    "    # let's look through our locus to find cut sites. We end at len-24, since we are looking thru 23-nts,  \n",
    "    # and because python includes the first index but not the last one\n",
    "    uniqueCount=0\n",
    "    degenerateCount=0\n",
    "    NAGcount=0 \n",
    "    \n",
    "    for i in range(len(targetSeq)-24):\n",
    "\n",
    "        # sub is an array containing all 23mers\n",
    "        sub=targetSeq[i:i+23]\n",
    "        \n",
    "        # first, cut sequences are N20NGG, cannot contain TTTTTT\n",
    "        if sub[21]==\"G\" and sub[22]=='G' and not('TTTTTT' in sub):\n",
    "           \n",
    "            # and also, if the last 12 bases exists followed by any NGG more than once, toss it.\n",
    "            # we define a CheckUnique function below which returns 1 if the input is found only once,\n",
    "            # returns 2 if it's found more than once, and 0 if it's not found at all. \n",
    "        \n",
    "            if (checkUnique(allRecords,sub[8:20]+\"AGG\")+checkUnique(allRecords,sub[8:20]+\"TGG\")+checkUnique(allRecords,sub[8:20]+\"CGG\")+checkUnique(allRecords,sub[8:20]+\"GGG\")) > 1:\n",
    "                break # If you find more than one instance of candidate N20 followed by any NGG, discard\n",
    "                \n",
    "            \n",
    "            # also, don't want any N9-20 followed by NAG\n",
    "            elif (checkUnique(allRecords,sub[8:20]+\"AAG\")+checkUnique(allRecords,sub[8:20]+\"TAG\")+checkUnique(allRecords,sub[8:20]+\"CAG\")+checkUnique(allRecords,sub[8:20]+\"GAG\")) > 0:\n",
    "                break # If you find any instance of candidate N20 followed by any NAG, discard\n",
    "                \n",
    "            \n",
    "            else:\n",
    "                # print the cut site that passes all tests\n",
    "                uniqueCount=uniqueCount+1\n",
    "                print (sub[0:20])\n",
    "    print(\"\\n\"+\"Done finding sequences. Good luck!\")  \n",
    "    print(\"degenerate, NAG counts: \")\n",
    "    print(degenerateCount)\n",
    "    print(NAGcount)\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "            \n",
    "            \n",
    "            \n",
    "def checkUnique(searchRecords, string):\n",
    "# Here we take in a string sequence and see if it's unique in the records (chromosomes)\n",
    "# defined earlier. If unique, returns 1; if not found, 0; and if found more than once, 2.\n",
    "        \n",
    "        nfound=0\n",
    "        \n",
    "        for record in searchRecords:\n",
    "            nfound=nfound+record.seq.count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            # note that we should look in the other strand, too    \n",
    "            nfound=nfound+record.seq.reverse_complement().count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            \n",
    "        if nfound == 0:\n",
    "            return 0\n",
    "            \n",
    "        if nfound == 1:\n",
    "            return 1\n",
    "            \n",
    "            \n",
    "findCutSites()\n",
    "                    \n",
    "                        \n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
