{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hi! Welcome to cutFinder. \n",
      "For questions or comments email leodespaux@gmail.com\n",
      "\n",
      "\n",
      "If you're using Python2, use quotes in your answers.\n",
      "I will only look at the sense strand of your target.\n",
      "\n",
      "\n",
      "I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL.\n",
      "Which one do you want? e.g., S288c: S288c\n",
      "\n",
      "Hold tight while I fetch that genome...\n",
      "\n",
      "I found the following genomic records: \n",
      "tpg|BK006935.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=I] [note=R64-1-1]\n",
      "tpg|BK006936.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=II] [note=R64-1-1]\n",
      "tpg|BK006937.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=III] [note=R64-1-1]\n",
      "tpg|BK006938.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IV] [note=R64-1-1]\n",
      "tpg|BK006939.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=V] [note=R64-1-1]\n",
      "tpg|BK006940.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VI] [note=R64-1-1]\n",
      "tpg|BK006941.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VII] [note=R64-1-1]\n",
      "tpg|BK006934.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VIII] [note=R64-1-1]\n",
      "tpg|BK006942.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IX] [note=R64-1-1]\n",
      "tpg|BK006943.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=X] [note=R64-1-1]\n",
      "tpg|BK006944.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XI] [note=R64-1-1]\n",
      "tpg|BK006945.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XII] [note=R64-1-1]\n",
      "tpg|BK006946.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIII] [note=R64-1-1]\n",
      "tpg|BK006947.3| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIV] [note=R64-1-1]\n",
      "tpg|BK006948.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XV] [note=R64-1-1]\n",
      "tpg|BK006949.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XVI] [note=R64-1-1]\n",
      "\n",
      "OK, what sequence do you want to find cut sites in: \n",
      "CTAAGAGGAGTACTTGGCAATCAGTTGATCAACAAGGGCAATGTATTCCTTTTCGGCATCTTCCTGGGATTTACCTTTTAAGTTTTCCCAGGCTTCCCACTTGTAGCGGTCCTTCATGTTGAAAATACCAGGCTTTTCCTTGTCGTTGTCACCTACAGTGGCTTGCTTGTACAGAGCATACAATTCTAATAATTCATCAGTGGAGGGCTTCGTTGGTAGCTCGTTGACGGCTTTAGCTTTTTCTTCGAATAATTGGGAAACCAT\n",
      "\n",
      "Found 1 instances of your target sequence in this genome.\n",
      "Now I'll find cut sites. I'll print them out as I find them so you know I'm working.\n",
      "\n",
      "TGGCAATCAGTTGATCAACA\n",
      "GGCAATCAGTTGATCAACAA\n",
      "AAGGGCAATGTATTCCTTTT\n",
      "ATTCCTTTTCGGCATCTTCC\n",
      "TTCCTTTTCGGCATCTTCCT\n",
      "TTTACCTTTTAAGTTTTCCC\n",
      "CCCAGGCTTCCCACTTGTAG\n",
      "TCCTTCATGTTGAAAATACC\n",
      "CTTGTCGTTGTCACCTACAG\n",
      "CAATTCTAATAATTCATCAG\n",
      "TTCTAATAATTCATCAGTGG\n",
      "TCTAATAATTCATCAGTGGA\n",
      "TCATCAGTGGAGGGCTTCGT\n",
      "CTTCGTTGGTAGCTCGTTGA\n",
      "TAGCTTTTTCTTCGAATAAT\n",
      "AGCTTTTTCTTCGAATAATT\n",
      "\n",
      "These are the cut sites I found, and their degenerate and NAG counts.\n",
      "The lower counts are better. If you want, also try the rcomplement of your target as input.\n",
      "\n",
      "\n",
      "[['TGGCAATCAGTTGATCAACA', 1, 0],\n",
      " ['GGCAATCAGTTGATCAACAA', 1, 1],\n",
      " ['AAGGGCAATGTATTCCTTTT', 2, 1],\n",
      " ['ATTCCTTTTCGGCATCTTCC', 1, 0],\n",
      " ['TTCCTTTTCGGCATCTTCCT', 1, 2],\n",
      " ['TTTACCTTTTAAGTTTTCCC', 2, 0],\n",
      " ['CCCAGGCTTCCCACTTGTAG', 1, 0],\n",
      " ['TCCTTCATGTTGAAAATACC', 1, 0],\n",
      " ['CTTGTCGTTGTCACCTACAG', 1, 0],\n",
      " ['CAATTCTAATAATTCATCAG', 1, 2],\n",
      " ['TTCTAATAATTCATCAGTGG', 1, 0],\n",
      " ['TCTAATAATTCATCAGTGGA', 1, 0],\n",
      " ['TCATCAGTGGAGGGCTTCGT', 1, 0],\n",
      " ['CTTCGTTGGTAGCTCGTTGA', 1, 0],\n",
      " ['TAGCTTTTTCTTCGAATAAT', 2, 1],\n",
      " ['AGCTTTTTCTTCGAATAATT', 1, 0]]\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "cutFinder\n",
    "by Leo d'Espaux <leodespaux@gmail.com>\n",
    "with help from William Zhuang, Kai Li\n",
    "\n",
    "Finds cut sites within a user-input sequence, then checks those candidates against\n",
    "chromosome files discarding sequences found more than once. \n",
    "Output is a list of candidate 20mer gRNAs in your target sequence.\n",
    "\n",
    "Details:\n",
    "*For now only looks in the sense strand of input, but checks both sense and antisense on the genome.\n",
    "*The program lists these sequences as it finds them. \n",
    "*Usually, we get a cut site every ~25nt.\n",
    "*Note your target can be a sequence that's not originally in the genome.\n",
    "\n",
    "Current as of 8/30/16\n",
    "\n",
    "'''\n",
    "\n",
    "# import libraries\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from pprint import pprint\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def findCutSites():\n",
    "    print(\"Hi! Welcome to cutFinder. \\nFor questions or comments email leodespaux@gmail.com\\n\\n\")\n",
    "    print(\"If you're using Python2, use quotes in your answers.\")\n",
    "    print(\"I will only look at the sense strand of your target.\\n\\n\")\n",
    "\n",
    "\n",
    "    print(\"I have the following genomes: S288c, CENpk, PO1f, CLIB, SAKL.\")\n",
    "    species=input(\"Which one do you want? e.g., S288c: \")\n",
    "    print(\"\")\n",
    "    print(\"Hold tight while I fetch that genome...\\n\")\n",
    "    \n",
    "    if species==\"PO1f\":\n",
    "        allRecords=[SeqIO.read(\"Yl_PO1f_A.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_B.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_C.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_D.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_E.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_F.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"CLIB\":\n",
    "        allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"SAKL\":\n",
    "        allRecords=[SeqIO.read(\"chromosomeA.fasta\", \"fasta\"), SeqIO.read(\"chromosomeB.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeC.fasta\", \"fasta\"), SeqIO.read(\"chromosomeD.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeE.fasta\", \"fasta\"), SeqIO.read(\"chromosomeF.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"S288c\":\n",
    "        allRecords=[SeqIO.read(\"Scer01.fasta\", \"fasta\"), SeqIO.read(\"Scer02.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer03.fasta\", \"fasta\"), SeqIO.read(\"Scer04.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer05.fasta\", \"fasta\"), SeqIO.read(\"Scer06.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer07.fasta\", \"fasta\"), SeqIO.read(\"Scer08.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer09.fasta\", \"fasta\"), SeqIO.read(\"Scer10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer11.fasta\", \"fasta\"), SeqIO.read(\"Scer12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer13.fasta\", \"fasta\"), SeqIO.read(\"Scer14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer15.fasta\", \"fasta\"), SeqIO.read(\"Scer16.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"CENpk\":\n",
    "        allRecords=[SeqIO.read(\"CENPK113-7D_CH1.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH2.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH3.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH4.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH5.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH6.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH7.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH8.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH9.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH11.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH13.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_CH15.fasta\", \"fasta\"), SeqIO.read(\"CENPK113-7D_CH16.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"CENPK113-7D_mitochondria.fasta\", \"fasta\")]\n",
    "        \n",
    "    print(\"I found the following genomic records: \")\n",
    "    for record in allRecords: # a record is a chromosome file\n",
    "        print( record.description)\n",
    "    print(\"\")\n",
    "\n",
    "    \n",
    "    \n",
    "    # ask for target and see if it's present and unique\n",
    "    print(\"OK, what sequence do you want to find cut sites in: \")\n",
    "    targetSeq=Seq(input().replace(\" \",\"\").upper()) #get rid of spaces and make all caps\n",
    "    \n",
    " \n",
    "    targetCount=checkUnique(allRecords, targetSeq)\n",
    "    print(\"\\nFound \"+str(targetCount)+\" instances of your target sequence in this genome.\")\n",
    "    print(\"Now I'll find cut sites. I'll print them out as I find them so you know I'm working.\\n\")\n",
    "    \n",
    "    #some sensible checks\n",
    "    #if targetCount==0:\n",
    "     #   print(\"Is this a heterologous gene? I'll still find cut sites.\")\n",
    "        \n",
    "    #elif targetCount>1:\n",
    "      #  print(\"There are multiple instances of your target sequence in this genome. I'll still find cut sites.\")\n",
    "    \n",
    "    #elif targetCount==1:\n",
    "      #  print(\"Your target sequence is found in the genome once.\")\n",
    "    \n",
    "\n",
    "    # make an empty list of cut sites\n",
    "    cutters=[]    \n",
    "    \n",
    "    # let's look through our locus to find cut sites. We end at len-24, since we are looking thru 23-nts,  \n",
    "    # and because python includes the first index but not the last one\n",
    "    \n",
    "    \n",
    "    for i in range(len(targetSeq)-24):\n",
    "\n",
    "        # sub is an array containing all 23mers\n",
    "        sub=targetSeq[i:i+23]\n",
    "        uniqueCount=0\n",
    "        degenerateCount=0\n",
    "        NAGcount=0 \n",
    "        # first, cut sequences are N20NGG, cannot contain TTTTTT since that terminates PolIII transcription\n",
    "        if sub[21]==\"G\" and sub[22]=='G' and not('TTTTTT' in sub):\n",
    "           \n",
    "            # First, we're going to look to find instances of N20 followed by any NGG. If there's more than once, then the cutter is NOT unique and we discard.\n",
    "        \n",
    "            if (checkUnique(allRecords,sub[1:20]+\"AGG\")+checkUnique(allRecords,sub[1:20]+\"TGG\")+checkUnique(allRecords,sub[1:20]+\"CGG\")+checkUnique(allRecords,sub[1:20]+\"GGG\")) > 1:\n",
    "                break # If you find more than one instance of candidate N20 followed by any NGG, discard\n",
    "            \n",
    "            else:\n",
    "                #degenerate count is the count of all instances of the seed sequence. Less the better.\n",
    "                degenerateCount=checkUnique(allRecords,sub[8:20]+\"AGG\")+checkUnique(allRecords,sub[8:20]+\"TGG\")+checkUnique(allRecords,sub[8:20]+\"CGG\")+checkUnique(allRecords,sub[8:20]+\"GGG\") \n",
    "                \n",
    "                #NAG count is the count of instances of the seed sequence followed by NAG, which can have off target effects. Less is better.\n",
    "                NAGcount=checkUnique(allRecords,sub[8:20]+\"AAG\")+checkUnique(allRecords,sub[8:20]+\"TAG\")+checkUnique(allRecords,sub[8:20]+\"CAG\")+checkUnique(allRecords,sub[8:20]+\"GAG\")\n",
    "                \n",
    "                #Then append and print.\n",
    "                cutters.append([str(sub[0:20]), degenerateCount, NAGcount])\n",
    "                print(sub[0:20])\n",
    "                \n",
    "    if not cutters:\n",
    "        print(\"\\nSorry didn't find any suitable cut sites. Consider starting again with the rcomplement of your target.\")\n",
    "    else:\n",
    "        print(\"\\nThese are the cut sites I found, and their degenerate and NAG counts.\")  \n",
    "        print(\"The lower counts are better. If you want, also try the rcomplement of your target as input.\\n\\n\")\n",
    "        pprint(cutters)\n",
    "\n",
    "    #print(\"degenerate, NAG counts: \")\n",
    "    #print(degenerateCount)\n",
    "    #print(NAGcount)\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "            \n",
    "            \n",
    "            \n",
    "def checkUnique(searchRecords, string):\n",
    "# Here we take in a string sequence and see if it's unique in the records (chromosomes)\n",
    "# defined earlier. If unique, returns 1; if not found, 0; and if found more than once, 2.\n",
    "        \n",
    "        nfound=0\n",
    "        \n",
    "        for record in searchRecords:\n",
    "            nfound=nfound+record.seq.count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            # note that we should look in the other strand, too    \n",
    "            nfound=nfound+record.seq.reverse_complement().count(str(string))\n",
    "            if nfound>1:\n",
    "                return 2\n",
    "                break\n",
    "            \n",
    "        if nfound == 0:\n",
    "            return 0\n",
    "            \n",
    "        if nfound == 1:\n",
    "            return 1\n",
    "            \n",
    "            \n",
    "findCutSites()\n",
    "                    \n",
    "                        \n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
