{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Hold tight while I fetch that genome...\n",
      "\n",
      "I found the following genomic records: \n",
      "tpg|BK006935.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=I] [note=R64-1-1]\n",
      "tpg|BK006936.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=II] [note=R64-1-1]\n",
      "tpg|BK006937.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=III] [note=R64-1-1]\n",
      "tpg|BK006938.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IV] [note=R64-1-1]\n",
      "tpg|BK006939.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=V] [note=R64-1-1]\n",
      "tpg|BK006940.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VI] [note=R64-1-1]\n",
      "tpg|BK006941.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VII] [note=R64-1-1]\n",
      "tpg|BK006934.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=VIII] [note=R64-1-1]\n",
      "tpg|BK006942.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=IX] [note=R64-1-1]\n",
      "tpg|BK006943.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=X] [note=R64-1-1]\n",
      "tpg|BK006944.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XI] [note=R64-1-1]\n",
      "tpg|BK006945.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XII] [note=R64-1-1]\n",
      "tpg|BK006946.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIII] [note=R64-1-1]\n",
      "tpg|BK006947.3| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XIV] [note=R64-1-1]\n",
      "tpg|BK006948.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XV] [note=R64-1-1]\n",
      "tpg|BK006949.2| [organism=Saccharomyces cerevisiae S288c] [strain=S288c] [moltype=genomic] [chromosome=XVI] [note=R64-1-1]\n",
      "\n",
      "OK, what sequence do you want to find cut sites in: \n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "cutFinder\n",
    "by Leo d'Espaux  \n",
    "\n",
    "contributions by: Xingkai Li\n",
    "\n",
    "Finds cut sites within a user-input sequence, having the chromosome files saved locally\n",
    "or retrieving from NCBI using entrez. The \"_Count\" version rather than discarding candidate cut sequences\n",
    "with more than one instance per chromosome, provides a list of all sequences and their Ncounts, and\n",
    "NAGcoutns. \n",
    "\n",
    "Current as of 5/2/16\n",
    "\n",
    "\n",
    "\n",
    "DIRECTIONS:\n",
    "Click on this text, press play above, and wait. If stuck, press the restart button, and repeat. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "# import libraries\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def findCutSites():\n",
    "\n",
    "    \n",
    "    # First, let's get the genome files and check that they load correctly \n",
    "    \n",
    "    #print(\"What species do you want, e.g. PO1f, s288c\")\n",
    "    species=\"s288c\" #input(\"Or type server to fetch the genome. \")\n",
    "    print(\"\")\n",
    "    print(\"Hold tight while I fetch that genome...\")\n",
    "    print(\"\")\n",
    "    \n",
    "    if species==\"server\":\n",
    "        allRecords=fetchNCBI()\n",
    "        \n",
    "    elif species==\"PO1f\":\n",
    "        allRecords=[SeqIO.read(\"Yl_PO1f_A.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_B.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_C.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_D.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Yl_PO1f_E.fasta\", \"fasta\"), SeqIO.read(\"Yl_PO1f_F.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"CLIB\":\n",
    "        allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "            SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "\n",
    "    elif species==\"SAKL\":\n",
    "        allRecords=[SeqIO.read(\"chromosomeA.fasta\", \"fasta\"), SeqIO.read(\"chromosomeB.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeC.fasta\", \"fasta\"), SeqIO.read(\"chromosomeD.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"chromosomeE.fasta\", \"fasta\"), SeqIO.read(\"chromosomeF.fasta\", \"fasta\")]\n",
    "        \n",
    "    elif species==\"s288c\":\n",
    "        allRecords=[SeqIO.read(\"Scer01.fasta\", \"fasta\"), SeqIO.read(\"Scer02.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer03.fasta\", \"fasta\"), SeqIO.read(\"Scer04.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer05.fasta\", \"fasta\"), SeqIO.read(\"Scer06.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer07.fasta\", \"fasta\"), SeqIO.read(\"Scer08.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer09.fasta\", \"fasta\"), SeqIO.read(\"Scer10.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer11.fasta\", \"fasta\"), SeqIO.read(\"Scer12.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer13.fasta\", \"fasta\"), SeqIO.read(\"Scer14.fasta\", \"fasta\"),\n",
    "                    SeqIO.read(\"Scer15.fasta\", \"fasta\"), SeqIO.read(\"Scer16.fasta\", \"fasta\"),]\n",
    "        \n",
    "    # Let's print the records to make sure they're good\n",
    "    print(\"I found the following genomic records: \")\n",
    "    for record in allRecords:\n",
    "        print( record.description)\n",
    "    print(\"\")\n",
    "\n",
    "    \n",
    "\n",
    "    # ask for target sequence and see if it's present and unique\n",
    "    print(\"OK, what sequence do you want to find cut sites in: \")\n",
    "    targetSeq=Seq(input().replace(\" \",\"\")) #also want to remove annoying spaces\n",
    "    \n",
    "    print(\"\\nLet me do some due diligence...\\n\")\n",
    "    targetCount=checkUnique(allRecords, targetSeq)\n",
    "    print(\"Found \"+str(targetCount)+\" instances of your target sequence in all the chromosomes (+ and - strands)\\n\")\n",
    "     \n",
    "    \n",
    "    # some sensible checks\n",
    "    if targetCount==0:\n",
    "        print(\"Sorry your sequence is not found in this genome.\")\n",
    "        exit()\n",
    "   \n",
    "\n",
    "    #Now let's find the cut sites we want\n",
    "    print(\"\\nAnd here is a list of the guides, instances of each found as N12-20NGG, and as N12-20NAG, in all chromosomes:\\n\")\n",
    " \n",
    "    #make an empty list\n",
    "    cutters=[]    \n",
    "    \n",
    "    #And now we'll run the sequence in the fwd direction, and then in the reverse\n",
    "    #And print any cut sites and Ninstances\n",
    "    findCandidates(targetSeq, allRecords, cutters)\n",
    "    findCandidates(targetSeq.reverse_complement(), allRecords, cutters)\n",
    "    \n",
    "    print(cutters)\n",
    "    \n",
    "    if cutters==[]:\n",
    "        print(\"Shit out of luck. No cutters.\")\n",
    "    \n",
    "\n",
    "def findCandidates(targetSeq, allRecords, cutters):\n",
    "    \n",
    "    # let's look through our locus to find cut sites. We end at len-24, since we are looking thru 23-nts,  \n",
    "    # and because python includes the first index but not the last one\n",
    "    \n",
    "    for i in range(len(targetSeq)-24):\n",
    "        \n",
    "        \n",
    "        \n",
    "        # And now run through candidate cut sequences, sub\n",
    "        sub=targetSeq[i:i+23]\n",
    "        \n",
    "        #make empty counters for each sub sequence\n",
    "        degenerateCount=0\n",
    "        NAGcount=0 \n",
    "        \n",
    "        # first, cut sequences are N20NGG, cannot contain TTTTTT\n",
    "        if sub[21]==\"G\" and sub[22]=='G' and not('TTTTTT' in sub):\n",
    "           \n",
    "            # Checkunique, defined below, returns 0, 1, or N, each being the number of instances of the sub sequence that matters\n",
    "            # which is the last 12 nt according to DiCarlo NAR 2013\n",
    "        \n",
    "            degenerateCount=checkUnique(allRecords,sub[8:20]+\"AGG\")+checkUnique(allRecords,sub[8:20]+\"TGG\")+checkUnique(allRecords,sub[8:20]+\"CGG\")+checkUnique(allRecords,sub[8:20]+\"GGG\")\n",
    "                \n",
    "            # also, don't want any N9-20 followed by NAG\n",
    "            NAGcount=checkUnique(allRecords,sub[8:20]+\"AAG\")+checkUnique(allRecords,sub[8:20]+\"TAG\")+checkUnique(allRecords,sub[8:20]+\"CAG\")+checkUnique(allRecords,sub[8:20]+\"GAG\")\n",
    "                \n",
    "            cutters.append([str(sub[0:20]), degenerateCount, NAGcount])\n",
    "                    \n",
    "  \n",
    "            \n",
    "            \n",
    "            \n",
    "def checkUnique(allRecords, string):\n",
    "# Here we take in a string sequence and see if it's unique in the records (chromosomes)\n",
    "# defined earlier. If unique, returns 1; if not found, 0; and if found more than once, returns number of instances\n",
    "        \n",
    "        nfound=0\n",
    "        \n",
    "        for record in allRecords: #each record is a chromosome file. We need to look at the reverse too.\n",
    "            nfound=nfound+record.seq.count(str(string))+record.seq.reverse_complement().count(str(string))\n",
    "            \n",
    "        if nfound>1:\n",
    "            return nfound\n",
    "            \n",
    "        elif nfound == 0:\n",
    "            return 0\n",
    "            \n",
    "        elif nfound == 1:\n",
    "            return 1\n",
    "            \n",
    "            \n",
    "findCutSites()\n",
    "                    \n",
    "                        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "   \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "Hold on... It's loading...\n",
    "\n",
    "\n",
    "''' \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
